File: Vol.12(2100-2199)/#2193 Window Pains/p2193.cpp
// 1835712 2009-04-17 00:06:00 Accepted  2173 C++ 0 184 呆滞的慢板 

// 模拟题，检查窗口覆盖是否有不合法的情况。
// 通过各个窗口的位置，构有向图 G[i][j] 如果 i 块在 j 块下面 G[i][j] 为真。
// 然后如果 G[i][j] 无环则合法，否则不合法。 

#include <iostream>
#include <string>
#include <vector>
using namespace std;

void tidy(string& x) {
    int j = 0;
    for(int i = 0; i < x.size(); ++i)
        if(x[i] != ' ')
            x[j++] = x[i];
    x.erase(j);
}

int main() {
    string x;
    string P[4];
    bool G[9][9];
    vector<int> H[4][4];
    for(int i = 0; i < 3; ++i) {
        for(int j = 0; j < 3; ++j) {
            H[i][j].push_back(i*3+j);
            H[i][j+1].push_back(i*3+j);
            H[i+1][j].push_back(i*3+j);
            H[i+1][j+1].push_back(i*3+j);
        }
    }
    while(getline(cin, x) && x == "START") {
        for(int i = 0; i < 4; ++i) {
            getline(cin, P[i]);
            tidy(P[i]);
        }
        getline(cin, x);
        memset(G, 0, sizeof(G));
        for(int i = 0; i < 4; ++i)
            for(int j = 0; j < 4; ++j)
                for(int k = 0; k < H[i][j].size(); ++k)
                    if(H[i][j][k] != P[i][j] - '1')
                        G[P[i][j] - '1'][H[i][j][k]] = true;
        for(int i = 0; i < 9; ++i)
            for(int j = 0; j < 9; ++j)
                for(int k = 0; k < 9; ++k)
                    if(G[j][i] && G[i][k])
                        G[j][k] = true;
        bool ans = false;
        for(int i = 0; i < 9; ++i)
            if(G[i][i] == true)
                ans = true;
        puts(
            ans ?
            "THESE WINDOWS ARE BROKEN" :
            "THESE WINDOWS ARE CLEAN"
        );
    }
}

File: 范例/Accpt/21xx/p2193.pas
PROGRAM p2193;

VAR
        i,j                                     :Integer;
        s                                       :String;
        map                                     :array[1..4,1..4]of Char;

FUNCTION Check(n:Integer):Boolean;
var
        p,q,i,j                                 :Integer;
begin
p:=(n+2) div 3;
q:=n mod 3;
if q=0 then q:=3;
for i:=p to p+1 do
        for j:=q to q+1 do
                if (map[i,j]<>'.')and(map[i,j]<>char(n+48)) then
                begin
                check:=false;
                exit;
                end;
check:=true;
for i:=p to p+1 do
        for j:=q to q+1 do map[i,j]:='.';
end;

FUNCTION Ans:Boolean;
var
        find                                    :Boolean;
        i,j                                     :Integer;
        done                                    :array[1..9]of Boolean;
begin
fillchar(done,sizeof(done),false);
repeat
        find:=false;
        for i:=1 to 9 do
                if not done[i] and (check(i)) then
                        begin
                        done[i]:=true;
                        find:=true;
                        end;
until not find;
Ans:=true;
for i:=1 to 4 do
        for j:=1 to 4 do
                if map[i,j]<>'.' then
                        begin
                        ans:=false;
                        exit;
                        end;
end;


BEGIN
readln(s);
while (s<>'ENDOFINPUT') do
        begin
        for i:=1 to 4 do
                begin
                read(map[i,1]);
                for j:=2 to 4 do read(map[i,j],map[i,j]);
                readln;
                end;
        if Ans then writeln('THESE WINDOWS ARE CLEAN')
                else writeln('THESE WINDOWS ARE BROKEN');
        readln;
        readln(s);
        end;
END.